### PR TITLE
Configmap Extended Lab

### PR DESCRIPTION
**Pull request by phil-pona**
This PR adds the extended lab on how to use configmaps, it's based on the kubernetes techlab

### CODE REVIEW

File: additional-labs/configmaps.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
+# Lab 10.4: ConfigMap
+
+ConfigMaps erlauben ähnlich wie das setzen von Umgebungsvariablen die Konfiguration für eine Applikation vom Image zu trennen und bei Laufzeit dem Pod zur Verfügung zu stellen um Applikationen innerhalb von Containern möglichst portabel zu halten.
```
Comment (bliemli) [inline additional-labs/configmaps.md]:
```suggestion
ConfigMaps werden dazu verwendet, die Konfiguration für eine Applikation vom Image zu trennen und bei Laufzeit dem Pod zur Verfügung zu stellen, ähnlich dem Setzen von Umgebungsvariablen. Dies erlaubt es, Applikationen innerhalb von Containern möglichst portabel zu halten.
```

File: additional-labs/configmaps.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
+# Lab 10.4: ConfigMap
+
+ConfigMaps erlauben ähnlich wie das setzen von Umgebungsvariablen die Konfiguration für eine Applikation vom Image zu trennen und bei Laufzeit dem Pod zur Verfügung zu stellen um Applikationen innerhalb von Containern möglichst portabel zu halten.
+
+In diesem Lab lernen Sie wie man ConfigMaps erstellt und entsprechend verwendet.
```
Comment (bliemli) [inline additional-labs/configmaps.md]:
```suggestion
In diesem Lab lernen Sie, wie man ConfigMaps erstellt und entsprechend verwendet.
```

File: additional-labs/configmaps.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+```
+
+
+## Java properties Files als ConfigMap
+
+Ein klassisches Beispiel für ConfigMaps sind Property Files bei Java Applikationen, welche in erster Linie nicht via Umgebungsvariablen konfiguriert werden können.
+
+Für dieses Beispiel verwenden wir das Spring Boot Beispiel aus [LAB 4](../labs/04_deploy_dockerimage.md), `[USER]-dockerimage`. **Tipp:** `oc project [USER]-dockerimage`
+
+Mit dem folgenden Befehl legen wir nun die erste ConfigMap auf Basis eines localen Files an:
```
Comment (bliemli) [inline additional-labs/configmaps.md]:
```suggestion
Mit dem folgenden Befehl legen wir nun die erste ConfigMap auf Basis eines lokalen Files an:
```

File: additional-labs/configmaps.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+javaconfiguration   1      7s
+```
+kann nun verifiziert werden ob die ConfigMap erfolgreich angelegt wurde.
+
+Oder mittels `$ oc get configmaps javaconfiguration -o json` kann der Inhalt angezeigt werden. 
+
+
+## Configmap in Pod zur Verfügung stellen
+
+Als nächstest wollen wir die ConfigMap im Pod verfügbar machen.
```
Comment (bliemli) [inline additional-labs/configmaps.md]:
```suggestion
Als nächstes wollen wir die ConfigMap im Pod verfügbar machen.
```

File: additional-labs/configmaps.md [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+Grundsätzlich gibt es dafür die folgenden Möglichkeiten: https://docs.openshift.com/container-platform/3.11/dev_guide/configmaps.html#consuming-configmap-in-pods
```
Comment (bliemli) [inline additional-labs/configmaps.md]:
```suggestion
Grundsätzlich gibt es dafür die folgenden Möglichkeiten, welche in der [offiziellen Dokumentation](https://docs.openshift.com/container-platform/3.9/dev_guide/configmaps.html#consuming-configmap-in-pods) genauer beschrieben werden:
```

File: additional-labs/configmaps.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+
+* ConfigMap Properties als Umgebungsvariablen im Deployment 
+* Commandline Arguments via Umgebungsvariablen
+* als Volumes in den Container gemountet
+
+Im Beispiel hier wollen wir, dass das File als File auf einem Volume liegt.
+
+dafür müssen wir entweder den Pod oder in unserem Fall das Deployment `oc edit dc example-spring-boot` bearbeiten.
```
Comment (bliemli) [inline additional-labs/configmaps.md]:
```suggestion
Hierfür müssen wir entweder den Pod oder in unserem Fall das Deployment mit `oc edit dc example-spring-boot` bearbeiten.
```

File: additional-labs/configmaps.md [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+Zu beachten gilt es den volumeMounts (`spec.template.spec.containers.volumeMounts`wie wird das Volume in den Container gemountet) und volumes (`spec.template.spec.volumes` welches Volume in unserem Fall die Configmap wird in den Container gemountet) Teil
```
Comment (bliemli) [inline additional-labs/configmaps.md]:
```suggestion
Zu beachten gilt es dabei den volumeMounts- (`spec.template.spec.containers.volumeMounts`: wie wird das Volume in den Container gemountet) sowie den volumes-Teil (`spec.template.spec.volumes`: welches Volume in unserem Fall die ConfigMap wird in den Container gemountet).
```

File: additional-labs/configmaps.md [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+      terminationGracePeriodSeconds: 30
+      volumes:
+      - configMap:
+          defaultMode: 420
+          name: javaconfiguration
+        name: config-volume
+
+```
+
+anschliessend kann im Container im file /etc/config/properties.properties auf die Werte zugegriffen werden.
```
Comment (bliemli) [inline additional-labs/configmaps.md]:
```suggestion
Anschliessend kann im Container im File `/etc/config/properties.properties` auf die Werte zugegriffen werden.
```

File: additional-labs/configmaps.md [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+```bash
+$ oc exec [POD] cat /etc/config/properties.properties
+key=appuio
+key2=openshift
+```
+
+Diese Property File kann nun so von der Java Applikation im Conainer gelesen und verwendet werden. Das Image bleibt dabei umgebungsneutral.
```
Comment (bliemli) [inline additional-labs/configmaps.md]:
```suggestion
Diese Property File kann nun so von der Java Applikation im Container gelesen und verwendet werden. Das Image bleibt dabei umgebungsneutral.
```

File: additional-labs/configmaps.md [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+
+## Aufgabe: LAB10.4.1 ConfigMap Data Sources
+
+Erstellen Sie jeweils eine ConfigMap und verwenden Sie dafür die verschiedenen Arten von Data Sources https://docs.openshift.com/container-platform/3.11/dev_guide/configmaps.html#consuming-configmap-in-pods
```
Comment (bliemli) [inline additional-labs/configmaps.md]:
```suggestion
Erstellen Sie jeweils eine ConfigMap und verwenden Sie dafür die verschiedenen Arten von [Data Sources](https://docs.openshift.com/container-platform/3.9/dev_guide/configmaps.html#consuming-configmap-in-pods).
```

**Review by bliemli — CHANGES_REQUESTED** [review]

**Comment by phil-pona** [comment]
Thanks for the review and the fixes

**Comment by bliemli** [comment]
Thank you for the PR!

**Pull request merged**